### DISCUSSION TITLE
请问openwrt下的openclash要怎么启用自动更换IP

### DISCUSSION THREAD

**Discussion by SamZig** [XIU2/CloudflareSpeedTest#305] [讨论]
请问openwrt下的openclash要怎么启用自动更换IP

**Comment by XIU2** [comment]
不清楚，我没有 OpenWrt 设备。。。因此我连你说的 openclash 是什么都不知道。。。

**Comment by accounti** [comment]
用自带的脚本修改hosts 
然后再执行 /etc/init.d/openclash restart

**Comment by issacqin** [comment]
我猜，你的问题是，如果使用工具找出了高速的Cloudflare的IP节点，如何更新到自己的v2ray这类节点上。

在这里我是这样做的
1. 申请一个[changeip.com](https://www.changeip.com/accounts/aff.php?aff=3799)的服务，免费搞几个DDNS域名。
2. 你的节点地址就用这几个DDNS域名。
3. 找出的Cloudflare高速IP就影射到这些域名上就行了，这方案我用了好几年了。

bat脚本例子：
call curl "https://nic.changeip.com/nic/update?cmd=update&u=%DDNS_UID%&p=%DDNS_PWD%&hostname=%node_Current.DDNS_Name%&ip=!lastip!"

DDNS_UID = changeip.com user id
DDNS_PWD = changeip.com password
node_Current.DDNS_Name = ddns name
lastip = Cloudflare IP address

**Comment by hoockoo** [comment]
我也有你这个想法，打算周末写个脚本，目前思路是：
1、openclash创建5个套了cf但都是指向1个vps的节点。

2、stop openclash后用作者的优选工具选出5个较快的cf ip。

3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）

4、重新启动openclash。

**Comment by issacqin** [comment]
> 我也有你这个想法，打算周末写个脚本，目前思路是： 1、openclash创建5个套了cf但都是指向1个vps的节点。
> 
> 2、stop openclash后用作者的优选工具选出5个较快的cf ip。
> 
> 3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）
> 
> 4、重新启动openclash。

可以看看issue里面其它主题，有一个scan ip的项目我用下来感觉可以，找出的ip都延迟很小。
另外，直接改地址多粗暴，免费的ddns更新不优雅很多？
最后，我比较openclash和shadowsockets+,发现cf这个应用场景中，后者稳定多了。

**Comment by hoockoo** [comment]
> > 我也有你这个想法，打算周末写个脚本，目前思路是： 1、openclash创建5个套了cf但都是指向1个vps的节点。
> > 2、stop openclash后用作者的优选工具选出5个较快的cf ip。
> > 3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）
> > 4、重新启动openclash。
> 
> 可以看看issue里面其它主题，有一个scan ip的项目我用下来感觉可以，找出的ip都延迟很小。 另外，直接改地址多粗暴，免费的ddns更新不优雅很多？ 最后，我比较openclash和shadowsockets+,发现cf这个应用场景中，后者稳定多了。

我主要是考虑到使用优选工具前需要先把openclash的‘**路由本机代理**’这个选项关掉才能正确地优选IP和openclash自身restart速度很慢才去直接stop它，虽然openclash的stop函数确实有点粗暴~
如果使用uci set openclash.config.router_self_proxy='0'再restart，估计20秒都过去了，还不如直接stop>优选IP>start，而重启passwall2大概只需3秒，SSRP我目前没在使用。
DDNS的方法自己使用确实方便，脚本也更好写，但如果脚本要分享给别人使用的话就需要别人也注册几个DDNS。
我目前遇到一个很严重的问题是作者大佬使用的分流测速地址并不稳定，考虑到每次优选IP使用的流量不少，我自己也没有多余的VPS搭测速服务器，导致目前很多IP测速为0，优选的准确度稍差。因为我目前的做法是先优选出新IP再和旧IP做速度对比，再选择是否更换节点IP，经常第一次优选出来的IP地址和旧地址做对比测试时就没速度了。

**Comment by issacqin** [comment]
> > > 我也有你这个想法，打算周末写个脚本，目前思路是： 1、openclash创建5个套了cf但都是指向1个vps的节点。
> > > 2、stop openclash后用作者的优选工具选出5个较快的cf ip。
> > > 3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）
> > > 4、重新启动openclash。
> > 
> > 
> > 可以看看issue里面其它主题，有一个scan ip的项目我用下来感觉可以，找出的ip都延迟很小。 另外，直接改地址多粗暴，免费的ddns更新不优雅很多？ 最后，我比较openclash和shadowsockets+,发现cf这个应用场景中，后者稳定多了。
> 
> 我主要是考虑到使用优选工具前需要先把openclash的‘**路由本机代理**’这个选项关掉才能正确地优选IP和openclash自身restart速度很慢才去直接stop它，虽然openclash的stop函数确实有点粗暴~ 如果使用uci set openclash.config.router_self_proxy='0'再restart，估计20秒都过去了，还不如直接stop>优选IP>start，而重启passwall2大概只需3秒，SSRP我目前没在使用。 DDNS的方法自己使用确实方便，脚本也更好写，但如果脚本要分享给别人使用的话就需要别人也注册几个DDNS。 我目前遇到一个很严重的问题是作者大佬使用的分流测速地址并不稳定，考虑到每次优选IP使用的流量不少，我自己也没有多余的VPS搭测速服务器，导致目前很多IP测速为0，优选的准确度稍差。因为我目前的做法是先优选出新IP再和旧IP做速度对比，再选择是否更换节点IP，经常第一次优选出来的IP地址和旧地址做对比测试时就没速度了。

测速和优选都有在没有外网的环境下做，很费时，而且，我认为并不是带宽越大越好，最好还是节点离用户侧更近，延迟更少，才越稳定，性能也好，所以，我最近都在用IP—scaner的结果集去测速，这样快很多，效果很好，选五个节点做负载平衡，一般4k稳定秒开。

**Comment by issacqin** [comment]
> > > 我也有你这个想法，打算周末写个脚本，目前思路是： 1、openclash创建5个套了cf但都是指向1个vps的节点。
> > > 2、stop openclash后用作者的优选工具选出5个较快的cf ip。
> > > 3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）
> > > 4、重新启动openclash。
> > 
> > 
> > 可以看看issue里面其它主题，有一个scan ip的项目我用下来感觉可以，找出的ip都延迟很小。 另外，直接改地址多粗暴，免费的ddns更新不优雅很多？ 最后，我比较openclash和shadowsockets+,发现cf这个应用场景中，后者稳定多了。
> 
> 我主要是考虑到使用优选工具前需要先把openclash的‘**路由本机代理**’这个选项关掉才能正确地优选IP和openclash自身restart速度很慢才去直接stop它，虽然openclash的stop函数确实有点粗暴~ 如果使用uci set openclash.config.router_self_proxy='0'再restart，估计20秒都过去了，还不如直接stop>优选IP>start，而重启passwall2大概只需3秒，SSRP我目前没在使用。 DDNS的方法自己使用确实方便，脚本也更好写，但如果脚本要分享给别人使用的话就需要别人也注册几个DDNS。 我目前遇到一个很严重的问题是作者大佬使用的分流测速地址并不稳定，考虑到每次优选IP使用的流量不少，我自己也没有多余的VPS搭测速服务器，导致目前很多IP测速为0，优选的准确度稍差。因为我目前的做法是先优选出新IP再和旧IP做速度对比，再选择是否更换节点IP，经常第一次优选出来的IP地址和旧地址做对比测试时就没速度了。

测速和优选都有在没有外网的环境下做，很费时，而且，我认为并不是带宽越大越好，最好还是节点离用户侧更近，延迟更少，才越稳定，性能也好，所以，我最近都在用IP—scaner的结果集去测速，这样快很多，效果很好，选五个节点做负载平衡，一般4k稳定秒开。

**Comment by issacqin** [comment]
> > > 我也有你这个想法，打算周末写个脚本，目前思路是： 1、openclash创建5个套了cf但都是指向1个vps的节点。
> > > 2、stop openclash后用作者的优选工具选出5个较快的cf ip。
> > > 3、修改/etc/openclash/config下在用的yaml文件，把proxies里面5个节点的server修改成第2步获取到的5个ip地址。（因为openclash无法用uci set修改节点）
> > > 4、重新启动openclash。
> > 
> > 
> > 可以看看issue里面其它主题，有一个scan ip的项目我用下来感觉可以，找出的ip都延迟很小。 另外，直接改地址多粗暴，免费的ddns更新不优雅很多？ 最后，我比较openclash和shadowsockets+,发现cf这个应用场景中，后者稳定多了。
> 
> 我主要是考虑到使用优选工具前需要先把openclash的‘**路由本机代理**’这个选项关掉才能正确地优选IP和openclash自身restart速度很慢才去直接stop它，虽然openclash的stop函数确实有点粗暴~ 如果使用uci set openclash.config.router_self_proxy='0'再restart，估计20秒都过去了，还不如直接stop>优选IP>start，而重启passwall2大概只需3秒，SSRP我目前没在使用。 DDNS的方法自己使用确实方便，脚本也更好写，但如果脚本要分享给别人使用的话就需要别人也注册几个DDNS。 我目前遇到一个很严重的问题是作者大佬使用的分流测速地址并不稳定，考虑到每次优选IP使用的流量不少，我自己也没有多余的VPS搭测速服务器，导致目前很多IP测速为0，优选的准确度稍差。因为我目前的做法是先优选出新IP再和旧IP做速度对比，再选择是否更换节点IP，经常第一次优选出来的IP地址和旧地址做对比测试时就没速度了。

测速和优选都有在没有外网的环境下做，很费时，而且，我认为并不是带宽越大越好，最好还是节点离用户侧更近，延迟更少，才越稳定，性能也好，所以，我最近都在用IP—scaner的结果集去测速，这样快很多，效果很好，选五个节点做负载平衡，一般4k稳定秒开。

**Comment by goophoo** [comment]
写了两个小脚本，可以完成你说的启用自动更换IP；
https://github.com/goophoo/auto-update-cf-cdn-for-openclash

**Comment by hoockoo** [comment]
> 写了两个小脚本，可以完成你说的启用自动更换IP； https://github.com/goophoo/auto-update-cf-cdn-for-openclash

提个小建议，测速前把openclash关掉，测速结果会更准确。或者uci set openclash.config.router_self_proxy='0'把’路由本机代理‘关掉，但可能是因为我用的是meta内核，貌似第二个方法并不生效。

**Comment by goophoo** [comment]
> > 写了两个小脚本，可以完成你说的启用自动更换IP； https://github.com/goophoo/auto-update-cf-cdn-for-openclash
> 
> 提个小建议，测速前把openclash关掉，测速结果会更准确。或者uci set openclash.config.router_self_proxy='0'把’路由本机代理‘关掉，但可能是因为我用的是meta内核，貌似第二个方法并不生效。

就是因为openclash干扰测速，所以crontab里的测速是在非openclash的路由器上完成的，再把最优结果ssh去openclash所在路由器。我的情况是把安装openclash的路由作为旁路由使用。

**Comment by 120318** [comment]
> > 写了两个小脚本，可以完成你说的启用自动更换IP； https://github.com/goophoo/auto-update-cf-cdn-for-openclash
> 
> 提个小建议，测速前把openclash关掉，测速结果会更准确。或者uci set openclash.config.router_self_proxy='0'把’路由本机代理‘关掉，但可能是因为我用的是meta内核，貌似第二个方法并不生效。

我不使用meta内核，这个关掉后reload配置，本机代理仍旧生效。

**Comment by hoockoo** [comment]
(https://github.com/XIU2/CloudflareSpeedTest/issues/156#issuecomment-1243671760)

**Comment by ICLXL** [comment]
可以把cloudflare 测速地址，加入规则配置 -DOMAIN,url,DIRECT
